src/server/sched_func.c

```diff
@@ -0,0 +1,288 @@
+ *
+ * Altair’s dual-license business model allows companies, individuals, and
+ * organizations to create proprietary derivative works of PBS Pro and distribute
+ * them - whether embedded or bundled with other software - under a commercial
+ * license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+/**
+ * @file    sched_func.c
+ *
+ *@brief
+ * 		sched_func.c - various functions dealing with schedulers
+ *
+ */
+#include <errno.h>
+#include <string.h>
+#include "scheduler.h"
+#include "log.h"
+#include "pbs_ifl.h"
+#include "pbs_db.h"
+#include "pbs_error.h"
+#include <memory.h>
+#ifndef PBS_MOM
+extern pbs_db_conn_t	*svr_db_conn;
+#endif
+
+
+/**
+ * @brief
+ * 		sched_alloc - allocate space for a pbs_sched structure and initialize
+ *		attributes to "unset"
+ *
+ * @param[in]	sched_name	- scheduler  name
+ *
+ * @return	pbs_sched *
+ * @retval	null	- space not available.
+ */
+pbs_sched *
+sched_alloc(char *sched_name) {
+	int i;
+	pbs_sched *psched;
+
+	psched = (pbs_sched *) malloc(sizeof(pbs_sched));
+
+	if (psched == (pbs_sched *) 0) {
+		log_err(errno, "sched_alloc", "no memory");
+		return NULL;
+	}
+	(void) memset((char *) psched, (int) 0, (size_t) sizeof(pbs_sched));
+	CLEAR_LINK(psched->sc_link);
+	strncpy(psched->sc_name, sched_name, PBS_MAXSCHEDNAME);
+	psched->sc_name[PBS_MAXSCHEDNAME] = '\0';
+
+	append_link(&svr_allscheds, &psched->sc_link, psched);
+
+	/* set the working attributes to "unspecified" */
+
+	for (i = 0; i < (int) SCHED_ATR_LAST; i++) {
+		clear_attr(&psched->sch_attr[i], &sched_attr_def[i]);
+	}
+
+	return (psched);
+}
+
+/**
+ * @brief find a scheduler
+ *
+ * @param[in]	sched_name - scheduler name
+ *
+ * @return	pbs_sched *
+ */
+
+pbs_sched * find_scheduler(char *sched_name) {
+	char *pc;
+	pbs_sched *psched;
+
+	if (!strcmp(sched_name, "1") || !strcmp(sched_name,"sched"))
+		return dflt_scheduler;
+	psched = (pbs_sched *) GET_NEXT(svr_allscheds);
+	while (psched != (pbs_sched *) 0) {
+		if (strcmp(sched_name, psched->sc_name) == 0)
+			break;
+		psched = (pbs_sched *) GET_NEXT(psched->sc_link);
+	}
+	return (psched);
+}
+
+/**
+ * @brief free sched structure
+ *
+ * @param[in]	psched	- The pointer to the sched to free
+ *
+ */
+void
+sched_free(pbs_sched *psched)
+{
+	int		 i;
+	attribute	*pattr;
+	attribute_def	*pdef;
+
+	/* remove any malloc working attribute space */
+
+	for (i=0; i < (int)SCHED_ATR_LAST; i++) {
+		pdef  = &sched_attr_def[i];
+		pattr = &psched->sch_attr[i];
+
+		pdef->at_free(pattr);
+	}
+
+	/* now free the main structure */
+	delete_link(&psched->sc_link);
+	(void)free((char*)psched);
+}
+
+/**
+ * @brief - purge scheduler from system
+ *
+ * @param[in]	psched	- The pointer to the delete to delete
+ *
+ * @return	error code
+ * @retval	0	- scheduler purged or scheduler not valid
```

Comment:
This isn't right, scheduler purged would mean everything was successful, scheduler not valid would mean that nothing deleted because scheduler wasn't valid. Both should have their own return values.